src/lib.rs

```diff
@@ -101,63 +95,170 @@ impl Router {
-    log::trace!("Loading bindle {}", name);
-    let cache_dir = module_cache_dir.join("_ASSETS");
-    let mut mods = runtime::bindle::bindle_to_modules(name, bindle_server, cache_dir)
-        .await
-        .map_err(|e| anyhow::anyhow!("Failed to turn Bindle into module configuration: {}", e))?;
-
-    mods.build_registry(cache_config_path, module_cache_dir, base_log_dir)
+    /// Sets the default host to use for virtual hosting. If this is already set in a
+    /// `ModulesConfig`, this will be ignored when the router is built
+    pub fn default_host(mut self, host: &str) -> Self {
+        self.default_host = host.to_owned();
+        self
+    }
+
+    pub fn global_env_vars(mut self, vars: HashMap<String, String>) -> Self {
+        self.global_env_vars = vars;
+        self
+    }
+
+    /// Build the router, loading the config from a toml file at the given path
+    pub async fn build_modules_toml(self, path: impl AsRef<Path>) -> anyhow::Result<Router> {
+        if !tokio::fs::metadata(&path)
+            .await
+            .map(|m| m.is_file())
+            .unwrap_or(false)
+        {
+            return Err(anyhow::anyhow!(
+                "no modules configuration file found at {}",
+                path.as_ref().display()
+            ));
+        }
+
+        let data = std::fs::read(path)?;
+        let mut modules: ModuleConfig = toml::from_slice(&data)?;
+
+        modules
+            .build_registry(
+                &self.cache_config_path,
+                &self.module_cache_dir,
+                &self.base_log_dir,
+            )
+            .await?;
+
+        Ok(self.build(modules))
+    }
+
+    /// Build the router, loading the config from a bindle with the given name fetched from the
+    /// provided server
+    pub async fn build_bindle(self, name: &str, bindle_server: &str) -> anyhow::Result<Router> {
+        log::info!("Loading bindle {}", name);
+        let cache_dir = self.module_cache_dir.join("_ASSETS");
+        let mut mods = runtime::bindle::bindle_to_modules(name, bindle_server, cache_dir)
+            .await
+            .map_err(|e| {
+                anyhow::anyhow!("Failed to turn Bindle into module configuration: {}", e)
+            })?;
+
+        mods.build_registry(
+            &self.cache_config_path,
+            &self.module_cache_dir,
+            &self.base_log_dir,
+        )
         .await?;
-    Ok(mods)
+
+        Ok(self.build(mods))
+    }
+
+    fn build(self, mut config: ModuleConfig) -> Router {
+        // Apply the global variables on top of the user defined ones. This will overwrite existing
+        // user variables and add any ones that don't exist
+        if !self.global_env_vars.is_empty() {
+            config.modules = config
+                .modules
+                .into_iter()
+                .map(|mut module| {
+                    match module.environment.as_mut() {
+                        Some(current) => current.extend(self.global_env_vars.clone()),
```

Comment:
Reason 1,413,712 why I :heart: rust: Real collections that can do things like `extend()`. (Did I mention that I was just doing reviews of Go code?)